### ISSUE TITLE
Ref is not compatible with React.Ref

### ISSUE THREAD

**Issue by kuoruan** [palantir/blueprint#4690]
#### Environment

- __Package version(s)__: 3.44.1
- __Operating System__: Linux
- __Browser name and version__: Chrome 90

#### Code Sandbox

[https://codesandbox.io/s/cocky-mclean-e5m4c](https://codesandbox.io/s/cocky-mclean-e5m4c)

#### Steps to reproduce

Use `react-hook-form`'s `Controller` to define a field. Pass it's ref to `inputRef`.

#### Actual behavior

```
Type 'Ref<any>' is not assignable to type 'IRef<HTMLInputElement> | undefined'.
  Type 'null' is not assignable to type 'IRef<HTMLInputElement> | undefined'.ts(2322)
```

#### Expected behavior

Type check pass.

#### Possible solution

https://github.com/palantir/blueprint/pull/4687

**Comment by adidahiya** [comment]
Cool, thanks for creating the sandbox to illustrate the problem. For now the easiest way to work around this would be to use `ref ?? undefined` instead of `ref` in your example.

Since Blueprint 4.0 is coming soon, I am inclined to avoid any more changes to ref types on 3.x (develop branch) and instead consider making changes on the `next` branch... either using `React.Ref` directly (note that that was basically the whole point of the types in `refs.ts`... to avoid importing `@types/react` as you did in #4687) or tweaking our local `Ref<T>` type to be more compatible with `React.Ref<T>`.

**Comment by adidahiya** [comment]
Should be fixed by #5507. We're able to move to `React.Ref` completely now that we don't support React 15.

**Comment by adidahiya** [comment]
Note for future reference, `MutableRefObject<T>` is sometimes incompatible with `Ref<T>`. This can usually be fixed with a more explicit `useRef` typedef. For example, see this sandbox with @types/react v16: https://codesandbox.io/p/sandbox/wild-hooks-m363kn?file=%2Fsrc%2FCoreExample.tsx%3A22%2C1

```ts
const goodRef = React.useRef<HTMLInputElement | null>(null);
const badRef = React.useRef<HTMLInputElement>();
```
